extensions/ql-vscode/src/view/common/telemetry.ts

```diff
@@ -0,0 +1,60 @@
+import { useEffect, useMemo, useRef } from "react";
+import { vscode } from "../vscode-api";
+
+/**
+ * A react effect that outputs telemetry events whenever the value changes.
+ *
+ * @param value Default value to pass to React.useState
+ * @param telemetryAction Name of the telemetry event to output
+ * @param options Extra optional arguments, including:
+ *   filterTelemetryOnValue: If provided, only output telemetry events when the
+ *       predicate returns true. If not provided always outputs telemetry.
+ *   debounceTimeout: If provided, will not output telemetry events for every change
+ *       but will wait until specified timeout happens with no new events ocurring.
+ */
+export function useTelemetryOnChange<S>(
+  value: S,
+  telemetryAction: string,
+  options?: {
+    filterTelemetryOnValue?: (value: S) => boolean;
+    debounceTimeoutMillis?: number;
+  },
+) {
+  const previousValue = useRef(value);
+  const filterTelemetryOnValue = options?.filterTelemetryOnValue;
+  const debounceTimeoutMillis = options?.debounceTimeoutMillis;
+
+  const sendTelemetryFunc = useMemo<() => void>(() => {
```

Comment:
So, the idea here is to ensure that if this effect gets invoked multiple times within the debounce period, earlier data will be overwritten with newer data?

It looks like there may be a problem. If `debounceTimeoutMillis` is set to (say) 100ms, and this is invoked every 99ms (forever), then it looks like the event will never be sent. Am I reading that right?

I _think_ what we want to do is avoid clearing the `timer`, and just make sure that each time we call this effect, then we update the `telemetryAction` to the latest action. This way, the event will always be sent at most every `debounceTimeoutMillis` and always with the latest `telemetryAction`.

Reply:
Maybe this is a theoretical problem and in practice we'll never hit it.

Reply:
I think the current behavior makes more sense. If the user doesn't stop typing, we wouldn't want to send the telemetry event yet because they haven't completed their action. I'd expect us to be interested in how many distinct searches a user does, rather than that they took 5 seconds to construct their search.

Alternatively, we could have something like a `maxWait` parameter [like the `use-debounce` library has](https://github.com/xnimorz/use-debounce#cancel-maxwait-and-memoization). If the debounce is 1 second and the max wait is 10 seconds, this would send the event 1 second after the last action, or after 10 seconds, whichever comes first.

Reply:
> So, the idea here is to ensure that if this effect gets invoked multiple times within the debounce period, earlier data will be overwritten with newer data?

Yes, the point here is that if the state value changes multiple times within the debounce period then it will not send a telemetry event, and it'll instead send it later once the value hasn't changed for the required period. The idea is that if a user interacts with a feature like the search bar that changes state on every keypress, we only want to send one (or at most a few) telemetry events instead of one for every keypress. Effectively we want it to count as one use of the feature rather than many uses.

We're using `useMemo` because we need the `sendTelemetryFunc` function to stay the same function object, so that we can keep track of the `NodeJS.Timeout` variable.

> It looks like there may be a problem. If `debounceTimeoutMillis` is set to (say) 100ms, and this is invoked every 99ms (forever), then it looks like the event will never be sent. Am I reading that right?

You're right that if the event keeps happening within the debounce period then the telemetry will never be sent. I think that's unlikely since that requires the user to make an infinite number of actions. These aren't responding to automated events, so it's unlikely that it'll happen like this where there's never a long enough gap between events. We're also only dealing with user-interaction telemetry so if it doesn't send it's also not that bad. In general I'm not concerned about it too much.

> I _think_ what we want to do is avoid clearing the `timer`, and just make sure that each time we call this effect, then we update the `telemetryAction` to the latest action. This way, the event will always be sent at most every `debounceTimeoutMillis` and always with the latest `telemetryAction`.

I'm afraid I'm not quite following the suggestion here. Could you show an example of what you mean, or explain further?

Reply:
> I'd expect us to be interested in how many distinct searches a user does, rather than that they took 5 seconds to construct their search.

I agree we ideally only want to send one telemetry event per search, regardless of if the user takes a while to type it out. That's the reason why it's got the debounce time of 1s which is quite long. I actually considered making it even longer, but settled on 1s. Sending a few events is not the end of the world, so long as we don't send one for every single keypress.

> Alternatively, we could have something like a maxWait parameter

That does make sense to me as an extra addition, but I'd wait to see if we need it first, so I'm hesitant to add it right now. It's certainly something we could look at in the future though if we think the telemetry we're getting isn't correct or good enough.

Reply:
Thanks for the explanation and discussion. I think this all makes sense to me right now.